File: cpp/0015_3Sum.cpp
/*
先排序，再递归，再用双指针，注意cur vector的处理
*/
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums,int k,int target) {
        vector<vector<int>> res;
        if(nums.empty())
            return res;

        sort(nums.begin(),nums.end());

        vector<int> cur;
        dfs_find(nums,nums.size(),0,target,k,cur,res);

        return res;
    }

    void dfs_find(vector<int> & nums, int len,int start,int target,int k,vector<int> &cur,vector<vector<int>> & res)
    {
        if(target<k*nums.front() || target>k*nums.back())
            return;

        if(k==2)
        {
            int l=start,r=len-1;

            while(l<r)
            {
                int cur_sum=nums[l]+nums[r];

                if(cur_sum==target)
                {
                    cur.push_back(nums[l]);
                    cur.push_back(nums[r]);

                    vector<int> cur_copy(cur);
                    res.push_back(cur_copy);
                    cur.pop_back();
                    cur.pop_back();

                    l++, r--;
                    while(l<r && nums[l]==nums[l-1])
                        l++;
                    while(l<r && nums[r]==nums[r+1])
                        r--;
                }

                else if(cur_sum<target)
                    l++
                else
                    r--
            }

            return;
        }

        for(int i=start;i<=len-k;i++)
        {
            int ele=nums[i];
            if(i>0 && nums[i-1]==ele)
                continue;

            cur.push_back(ele);
            dfs_find(nums,len,i+1,target-ele,k-1,cur,res);
            cur.pop_back();
        }
    }
};

File: python/0015_3Sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res=[]
        nums.sort()#从小到大

        for i in range(len(nums)-2):
            if nums[i] > 0:
                break
            if i>0 and nums[i] ==nums[i-1]:
                continue

            target=nums[i]
            left,right=i+1,len(nums)-1
            while right>left:
                if target+nums[left]+nums[right]==0:
                    res.append([target,nums[left],nums[right]])
                    while left < right and nums[left] == nums[left+1]:
                        left += 1
                    while left < right and nums[right] == nums[right-1]:
                        right -= 1
                    left += 1; right -= 1
                elif target+nums[left]+nums[right]<0:
                    left+=1
                else:
                    right-=1

        return res


